[inductor] Fix an AOTInductor missing output issue

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #105496

Summary: When an output buffer is reused instead of directly referring to the passed-in output, we need to explictly make a copy

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @ngimel @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov

desertfire committed:
[inductor] Fix an AOTInductor missing output issue

Summary: When an output buffer is reused instead of directly referring to the passed-in output, we need to explictly make a copy

[ghstack-poisoned]

jansel commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1058,6 +1058,12 @@ def define_kernel(
         # Output tensors are allocated by the AOT runtime.
         if V.graph.aot_mode:
+            for idx, output in enumerate(V.graph.graph_outputs):
+                if output.get_name() in self.reuses:
```
It seems better to just make `can_reuse()` return False for outputs rather than adding a copy.

desertfire replied:
Ok, I gave that a try and it revealed a subtle issue. Because we take a two-pass codegen, in the first pass we generate the python code the reuse will happen (because there is no passed-in output tensors). In the second pass, `can_reuse()` returns false, and then the generated triton kernel will be different (with different number of args).

I can fix that by introducing some state so that when WrapperCodeGen generates code, it knows it's in an aot mode, but I am afraid that will make more and more aot logic bleed into the base WrapperCodeGen. For the time being, can we keep the `copy_` here?

jansel replied:
I'll approve to unblock, but I think we should figure out how not to do a copy here.  The copy will slow down AotAutograd.

jansel reviewed (changes requested):
Add a test for this.

desertfire committed:
Update on "[inductor] Fix an AOTInductor missing output issue"


Summary: When an output buffer is reused instead of directly referring to the passed-in output, we need to explictly make a copy

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy ngimel yf225 chenyang78 kadeng muchulee8 aakhundov

[ghstack-poisoned]

desertfire commented on torch/_inductor/codegen/wrapper.py:
```diff
@@ -1194,31 +1200,34 @@ def codegen_tensor_option(self, device, dtype):
         cpp_device = self.codegen_device(device)
         return f"at::TensorOptions({cpp_device}).dtype({DTYPE_TO_ATEN[dtype]}))"
 
+    def codegen_allocation(self, buffer):
```
This was introduced to address another subtle issue. `AllocateLine` may insert  a `ReuseLine` when running its planning. So instead of inserting `AllocateLine` here, we can just write out the outputs assignment.

desertfire commented:
https://github.com/pytorch/pytorch/issues/105633

desertfire commented:
@pytorchbot merge
